File: ut/java/myserver.java
import java.util.*;
import java.io.*;
import RemoteFilesBindings.*;

public class myserver {
	public static class MyFile implements RemoteFilesBindings.IFile {
		protected String _filename;
		protected OutputStream fout;
		protected InputStream fin;

		public MyFile(String filename, String mode) throws Exception {
			_filename = filename;
			if (mode.equals("r")) {
				fin = new FileInputStream(filename);
			} else if (mode.equals("w")) {
				fout = new FileOutputStream(filename);
			} else {
				throw new RemoteFilesBindings.UnderlyingIOError(
						"invalid mode: " + mode,
						RemoteFilesBindings.Errno.EFAULT);
			}
		}

		public String get_filename() throws Exception {
			return _filename;
		}

		public RemoteFilesBindings.StatRes stat() throws Exception {
			return new RemoteFilesBindings.StatRes(new Integer(15),
					new Integer(12345), new Integer(17772), new Integer(1001),
					new Integer(1002), new Date(278346812),
					new Date(286621112), new Date(181775244));
		}

		public byte[] read(Integer count) throws Exception {
			byte[] buf = new byte[count];
			int size = fin.read(buf, 0, buf.length);
			if (size < 0) {
				size = 0; // eof - return empty array
			}
			return Arrays.copyOf(buf, size);
		}

		public void write(byte[] data) throws Exception {
			fout.write(data, 0, data.length);
		}

		public void close() throws Exception {
			if (fin != null) {
				fin.close();
			}
			if (fout != null) {
				fout.close();
			}
		}

		public void flush() throws Exception {
			throw new RemoteFilesBindings.UnderlyingIOError("cannot flush",
					RemoteFilesBindings.Errno.EFAULT);
		}
	}

	public static class MyHandler implements RemoteFilesBindings.IHandler {
		public void copy(RemoteFilesBindings.IFile src,
				RemoteFilesBindings.IFile dst) throws Exception {
			System.out.println("@copy: src= " + src + ", dst = " + dst);
			byte[] buf;
			while (true) {
				buf = src.read(10000);
				if (buf.length == 0)
					break;
				dst.write(buf);
			}
		}

		public RemoteFilesBindings.IFile open(String filename, String mode)
				throws Exception {
			System.out.println("@open: filename = '" + filename + "', mode = '"
					+ mode + "'");
			return new MyFile(filename, mode);
		}

		public List<String> pathToList(RemoteFilesBindings.IPath path,
				List<RemoteFilesBindings.IFile> spam,
				List<RemoteFilesBindings.IPath> bacon,
				List<RemoteFilesBindings.Moshe> eggs,
				List<RemoteFilesBindings.IPath> maps) throws Exception {
			List<String> arr = new ArrayList<String>();
			arr.add("hello");
			arr.add("world");
			return arr;
		}

	}

	public static void main(String[] args) {
		try {
			agnos.Servers.SimpleServer server = new agnos.Servers.SimpleServer(
					new RemoteFilesBindings.Processor(new MyHandler()),
					new agnos.Transports.SocketTransportFactory("localhost",
							17732));

			server.serve();
		} catch (Exception ex) {
			ex.printStackTrace();
		}
	}
}

File: ut/python/myserver.py
import sys
sys.path.append("..")

import os
import errno
from datetime import datetime

from agnos import SocketTransportFactory 
from agnos.servers import SimpleServer
import RemoteFiles_bindings as RemoteFiles


class FileImp(object):
    def __init__(self, filename, mode):
        self._file = open(filename, mode)
        self.filename = filename
    
    def read(self, count):
        try:
            return self._file.read(count)
        except IOError, ex:
            raise RemoteFiles.UnderlyingIOError(ex.strerror, RemoteFiles.Errno.get_by_name(errno.errorcode[ex.errno]))
    def write(self, data):
        try:
            return self._file.write(data)
        except IOError, ex:
            raise RemoteFiles.UnderlyingIOError(ex.strerror, RemoteFiles.Errno.get_by_name(errno.errorcode[ex.errno]))
    def close(self):
        self._file.close()
    def flush(self):
        try:
            self._file.flush()
        except IOError, ex:
            raise RemoteFiles.UnderlyingIOError(ex.strerror, RemoteFiles.Errno.get_by_name(errno.errorcode[ex.errno]))
    def stat(self):
        try:
            res = os.fstat(self._file.fileno())
        except IOError, ex:
            raise RemoteFiles.UnderlyingIOError(ex.strerror, RemoteFiles.Errno.get_by_name(errno.errorcode[ex.errno]))
        
        return RemoteFiles.StatRes(inode = res.st_ino, mode = res.st_mode,
            size = res.st_size, uid = res.st_uid, gid = res.st_gid,
            atime = datetime.fromtimestamp(res.st_atime),
            ctime = datetime.fromtimestamp(res.st_mtime),
            mtime = datetime.fromtimestamp(res.st_ctime))


class Handler(RemoteFiles.IHandler):
    def open(self, filename, mode):
        return FileImp(filename, mode)
    
    def copy(self, src, dst):
        while True:
            buf = src.read(16000)
            if not buf:
                break
            dst.write(buf)
        dst.flush()


if __name__ == "__main__":
    s = SimpleServer(RemoteFiles.Processor(Handler()), SocketTransportFactory(17731))
    try:
        s.serve()
    except KeyboardInterrupt:
        pass

